core/domain/stats_jobs_one_off.py

```diff
@@ -16,16 +16,524 @@
+            yield (item.exploration_id, {
+                'event_type': feconf.EVENT_TYPE_COMPLETE_EXPLORATION,
+                'version': item.exploration_version,
+                'session_id': item.session_id
+            })
+
+        elif isinstance(item, stats_models.StateHitEventLogEntryModel):
+            value = {
+                'event_type': feconf.EVENT_TYPE_STATE_HIT,
+                'version': item.exploration_version,
+                'state_name': item.state_name,
+                'session_id': item.session_id,
+                'created_on': utils.get_time_in_millisecs(item.created_on)
+            }
+            yield (item.exploration_id, value)
+
+        else:
+            total_answers_count = len(item.submitted_answer_list)
+            useful_feedback_count = 0
+            for answer in item.submitted_answer_list:
+                if answer['classification_categorization'] != (
+                        exp_domain.DEFAULT_OUTCOME_CLASSIFICATION):
+                    useful_feedback_count += 1
+            value = {
+                'event_type': GenerateV1StatisticsJob.EVENT_TYPE_STATE_ANSWERS,
+                'version': item.exploration_version,
+                'state_name': item.state_name,
+                'total_answers_count': total_answers_count,
+                'useful_feedback_count': useful_feedback_count
+            }
+            yield (item.exploration_id, value)
+
+    @staticmethod
+    def reduce(exp_id, stringified_values):
+        num_starts = 0
+        num_completions = 0
+        num_actual_starts = 0
+
+        try:
+            exploration = exp_services.get_exploration_by_id(exp_id)
+        except Exception as e:
+            logging.error(e)
+            return
+
+        latest_exp_version = exploration.version
+
+        version_numbers = range(1, latest_exp_version + 1)
+        explorations_by_version = (
+            exp_services.get_multiple_explorations_by_version(
+                exp_id, version_numbers))
+        exploration_stats_by_version = (
+            stats_services.get_multiple_exploration_stats_by_version(
+                exp_id, version_numbers))
+
+        # The list of exploration start counts mapped by version.
+        start_exploration_counts_by_version = collections.defaultdict(int)
+        # The list of exploration complete counts mapped by version.
+        complete_exploration_counts_by_version = collections.defaultdict(
+            lambda: 0)
+        # The dict of state hit counts mapped by version.
+        state_hit_counts_by_version = collections.defaultdict(
+            lambda: collections.defaultdict(lambda: {
+                'total_hit_count': 0,
+                'first_hit_count': 0
+            }))
+        # The dict of state answer counts mapped by version.
+        state_answer_counts_by_version = collections.defaultdict(
+            lambda: collections.defaultdict(lambda: {}))
```

Comment:
I'm not sure I understand this; why is the inner `lambda: {}` not just `int`?

Reply:
@seanlip Because the inner state_answers_counts are stored in a dict with two keys - total_answers_count and useful_feedback_count.

Reply:
As discussed, specify these keys explicitly and aggregate, similar to state_hit_counts_by_version.

Reply:
Done.